HeterogeneousCore/SonicTriton/src/TritonData.cc

```diff
@@ -1,38 +1,104 @@
 #include "HeterogeneousCore/SonicTriton/interface/TritonData.h"
 #include "HeterogeneousCore/SonicTriton/interface/triton_utils.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
+
 #include "model_config.pb.h"
 
 #include <cstring>
+#include <sstream>
 
 namespace ni = nvidia::inferenceserver;
 namespace nic = ni::client;
 
 namespace nvidia {
   namespace inferenceserver {
-    //in librequest.so, but corresponding header src/core/model_config.h not available
-    size_t GetDataTypeByteSize(const DataType dtype);
+    //in libgrpcclient.so, but corresponding header src/core/model_config.h not available
+    size_t GetDataTypeByteSize(const inference::DataType dtype);
+    inference::DataType ProtocolStringToDataType(const std::string& dtype);
   }  // namespace inferenceserver
 }  // namespace nvidia
 
+//dims: kept constant, represents config.pbtxt parameters of model (converted from google::protobuf::RepeatedField to vector)
+//fullShape: if batching is enabled, first entry is batch size; values can be modified
+//shape: view into fullShape, excluding batch size entry
 template <typename IO>
-TritonData<IO>::TritonData(const std::string& name, std::shared_ptr<IO> data)
-    : name_(name), data_(std::move(data)), batchSize_(0) {
-  //convert google::protobuf::RepeatedField to vector
-  const auto& dimsTmp = data_->Dims();
-  dims_.assign(dimsTmp.begin(), dimsTmp.end());
-
-  //check if variable dimensions
-  variableDims_ = anyNeg(dims_);
-  if (variableDims_)
-    productDims_ = -1;
-  else
-    productDims_ = dimProduct(dims_);
-
-  dtype_ = data_->DType();
-  dname_ = ni::DataType_Name(dtype_);
-  //get byte size for input conversion
-  byteSize_ = ni::GetDataTypeByteSize(dtype_);
+TritonData<IO>::TritonData(const std::string& name, const TritonData<IO>::TensorMetadata& model_info, bool noBatch)
+    : name_(name),
+      dims_(model_info.shape().begin(), model_info.shape().end()),
+      noBatch_(noBatch),
+      batchSize_(0),
+      fullShape_(dims_),
+      shape_(&*(fullShape_.begin() + (noBatch_ ? 0 : 1)), &*(fullShape_.end())),
```

Comment:
De-referencing `std::vector<T>::end()` is undefined behavior.

If the type of `ShapeView` should always correspond to the type of `fullShape_`, how about
```cpp
using ShapeType = std::vector<int64_t>;
using ShapeView = edm::Span<ShapeType::const_iterator>;
...
ShapeType fullShape_;
```
and here
```suggestion
      shape_(fullShape_.begin() + (noBatch_ ? 0 : 1), fullShape_.end()),
```
? That would work even if the `vector::iterator` would be something else than a pointer.